tools/functional-tester/etcd-tester/lease_stresser.go

```diff
@@ -0,0 +1,373 @@
+// Copyright 2015 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+)
+
+func init() {
+	grpclog.SetLogger(plog)
+}
+
+// time to live for lease
+const TTL = 10
+
+type LeaseStresser interface {
+	// Stress starts to stress lease on the etcd cluster
+	Stress() error
+	// Cancel cancels the lease stress test on the etcd cluster
+	Cancel()
+	// Report reports the success and failure of the lease stress test
+	Report() (success int, failure int)
+
+	InvariantCheck(ctx context.Context) error
+
+	Setup() error
+}
+
+type leaseStresser struct {
+	Endpoint string
+	cancel   func()
+	conn     *grpc.ClientConn
+	id       int
+	kvc      pb.KVClient
+	lc       pb.LeaseClient
+
+	success           int
+	failure           int
+	numOfLeases       int
+	numOfKeysPerLease int
+
+	aliveLeases   map[int64]struct{}
+	revokedLeases map[int64]struct{}
+
+	wg *sync.WaitGroup
+	mu sync.Mutex
+}
+type leaseStresserBuilder func(m *member, i int) LeaseStresser
+
+func newLeaseStressBuilder(m int, n int) leaseStresserBuilder {
+	return func(mem *member, i int) LeaseStresser {
+		return &leaseStresser{
+			Endpoint:          mem.grpcAddr(),
+			numOfLeases:       m,
+			numOfKeysPerLease: n,
+			id:                i,
+		}
+	}
+}
+
+func (ls *leaseStresser) Setup() error {
+	if ls.numOfLeases == 0 {
+		panic("expect numOfLeases to be set")
+	}
+
+	if ls.numOfKeysPerLease == 0 {
+		panic("expect numOfKeysPerLease to be set")
+	}
+
+	conn, err := grpc.Dial(ls.Endpoint, grpc.WithInsecure())
+	if err != nil {
+		return fmt.Errorf("%v (%s)", err, ls.Endpoint)
+	}
+	ls.kvc = pb.NewKVClient(conn)
+	ls.lc = pb.NewLeaseClient(conn)
+	ls.conn = conn
+	ls.aliveLeases = make(map[int64]struct{})
+	ls.revokedLeases = make(map[int64]struct{})
+
+	return nil
+}
+
+func (ls *leaseStresser) Stress() error {
+	plog.Printf("lease Stresser %v starting ...", ls)
+	ctx, cancel := context.WithCancel(context.Background())
+	ls.cancel = cancel
+	ls.run(ctx)
+	return nil
+}
+
+func (ls *leaseStresser) run(ctx context.Context) {
+
```

Comment:
remove empty line here. we do not usually put any empty lines between func declaration and the implementation in go.